File: 백준/Silver/24480. 알고리즘 수업 － 깊이 우선 탐색 2/알고리즘 수업 － 깊이 우선 탐색 2.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline
def dfs(d):
    global cnt
    visited[d] = True
    answer[d] = cnt
    graph[d].sort(reverse=True)
    for i in graph[d]:
        if not visited[i]:
            cnt += 1
            dfs(i)

N,M,R = map(int,input().split())

graph = [[]for i in range(N+1)]
visited = [False] * (N+1)
answer = [0] * (N+1)
cnt = 1
for i in range(M):
    u,v = map(int,input().split())
    graph[u].append(v)
    graph[v].append(u)

dfs(R)

for i in answer[1:]:
    print(i)

File: 백준/Silver/24480. 알고리즘 수업 － 깊이 우선 탐색 2/알고리즘 수업 － 깊이 우선 탐색 2.cc
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

vector<int> graph[100001];
bool vst[100001];
int result[100001];
int cnt = 0;
bool cmp(int a, int b) {
    return a > b;
}
void dfs(int node) {
    sort(graph[node].begin(),graph[node].end(),cmp);
    vst[node] = true;
    cnt++;
    result[node] = cnt;
    for (int i = 0; i < graph[node].size(); i++) {
        int inde=graph[node][i];
        if (!vst[inde]) {
            dfs(inde);
        }
    }
}

int main(void) {
    int n, m, r;
    cin >> n >> m >> r;
    for (int i = 0; i < m; i++) {
        int a,b;
        cin >> a >> b;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }
    dfs(r);
    for (int i = 1; i <= n; i++) {
        cout << result[i] << "\n";
    }
}
